feature: MySQL DDL Statements

iobruno opened the pull request:
## Summary

- Add MySQL DDLs (File: `mysql.ddl`)
  - Changes table name from `groups` to `artifactgroups` for MySQL, since it's a registered keyword
  - Add `GetNextSequenceValue` store procedure since MySQL doesn't have an equivalent to Postgres' RETURNING or SQLServer's OUTPUT
  - Dropped IDX_versions_4 on versions as MySQL can't index VARCHAR(1024)
  - Dropped IDX_props_2 on properties as MySQL can't index VARCHAR(1024)
- Configured and fixed several testContainers for MySQL
  - Fix pattern to identify PK violation on MySQL
- Overlay config for storage/mysql
- Registers "mysql" jdbc driver for eligiblity for nextSequenceValue on AbstractSqlRegistryStorage
- Add and Registers MySqlRegistryStorage for Maven


Note*⚠️ A few compromises had to be made to make it compatible with MySQL:

- On artifacts table:
  - Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512) due to the Primary Key exceeding the MySQL limit of 3072 bytes
- On rules table:
  - Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512) due to the Primary Key exceeding the MySQL limit of 3072 bytes
- On versions table:
  - Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512) due to the Primary Key exceeding the MySQL limit of 3072 bytes
  - Reduced version to VARCHAR(128) from VARCHAR(256) due to the constraint UQ_versions_1 exceeding the MySQL limit of 3072 bytes
  - Reduced description to VARCHAR(512) from VARCHAR(1024) due to the IDX_versions_4 exceeding the MySQL limit of 3072 bytes
- On properties table:
  - Reduced pvalue to VARCHAR(512) from VARCHAR(1024) due to the IDX_props_2 exceeding the MySQL limit of 3072 bytes

iobruno committed:
Add mysql.ddl file

* A few compromises had to be made to make it compatible with MySQL

* On artifacts table:

  * Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512)
    due to the Primary Key exceeding the MySQL limit of 3072 bytes

* On rules table:

  * Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512)
    due to the Primary Key exceeding the MySQL limit of 3072 bytes

* On versions table:

  * Reduced groupId and artifactId to VARCHAR(256) from VARCHAR(512)
    due to the Primary Key exceeding the MySQL limit of 3072 bytes

  * Reduced version to VARCHAR(128) from VARCHAR(256) due to the
    constraint UQ_versions_1 exceeding the MySQL limit of 3072 bytes

  * Reduced description to VARCHAR(512) from VARCHAR(1024) due to the
    IDX_versions_4 exceeding the MySQL limit of 3072 bytes

* On properties table:

  * Reduced pvalue to VARCHAR(512) from VARCHAR(1024) due to the
    IDX_props_2 exceeding the MySQL limit of 3072 bytes

iobruno committed:
Add MySQLStatements

* Include MySQL-specific queries

* Registers MySQLSqlStatements on SqlStatementsProvider

* Registers "mysql" jdbc driver for eligiblity for
  nextSequenceValue on AbstractSqlRegistryStorage

* TODO: Create a Stored Procedure for getNextSequenceValue()

iobruno committed:
Overlay config for storage/mysql build

iobruno committed:
Add storage/MySqlRegistryStorage mvn subproject

* Implement MySqlStorage and StorageProvider based on the
  MS SQL Server implementation

* Implement integration tests with MySQL w/ testcontainers

* Include quarkus-mysql driver and testcontainers for mysql
  on maven project

iobruno committed:
Register mysql profile for maven

* Register mysql build profile (-Pmysql) for the
  module storage/mysql

gr8routdoors commented on storage/mysql/src/test/java/io/apicurio/registry/storage/util/MySqlEmbeddedTestResource.java:
```diff
@@ -0,0 +1,91 @@
+
+        if ("mas".equals(currentEnv)) {
+            Map<String, String> props = new HashMap<>();
+            props.put("quarkus.datasource.jdbc.url", "jdbc:sqlserver://mysql;");
```
This is incorrect

iobruno replied:
Missed the last push

gr8routdoors commented on storage/mysql/src/test/java/io/apicurio/registry/storage/util/MySqlEmbeddedTestResource.java:
```diff
@@ -0,0 +1,91 @@
+        if (database != null) {
+            database.close();
+        }
+        System.out.println("[MsSqlEmbeddedTestResource] MS SQLServer test resource stopped.");
```
This should be changed

iobruno replied:
Forgot to commit the last changes on this. But it's fixed now.

iobruno committed:
Configure testContainers for MySQL

gr8routdoors commented on app/src/main/java/io/apicurio/registry/storage/impl/sql/MySQLSqlStatements.java:
```diff
@@ -0,0 +1,122 @@
+     */
+    @Override
+    public String getNextSequenceValue() {
+        //TODO: Create a Stored Procedure to return the VALUE after the UPSERT
```
Are you sure you need a stored procedure here?  Seems overkill.

iobruno replied:
I've got to test this some more, I took the MySQL and SQLServer implementation as the basis and MySQL doesn't have the equivalent. For instance:

**Postgres**:
```
    @Override
    public String getNextSequenceValue() {
        return "INSERT INTO sequences (tenantId, name, value) VALUES (?, ?, 1) ON CONFLICT (tenantId, name) DO UPDATE SET value = sequences.value + 1 RETURNING value";
    }
```  

**SQLServer**:
```
    @Override
    public String getNextSequenceValue() {
        return String.join(" ",
                "MERGE INTO sequences AS target",
                "USING (VALUES  (?, ?)) AS source (tenantId, name)",
                "ON (target.tenantId = source.tenantId AND target.name = source.name)",
                "WHEN MATCHED THEN",
                    "UPDATE SET value = target.value + 1",
                "WHEN NOT MATCHED THEN",
                    "INSERT (tenantId, name, value)",
                    "VALUES (source.tenantId, source.name, 1)",
                "OUTPUT INSERTED.value;");
    }
```

iobruno committed:
Overriding CommonSqlStatements on groups table

* In MySQL, 'groups' is a reserved keyword.
  To query the table, it must be enclosed in backticks: `groups`

iobruno committed:
Update tbl 'groups' to 'artifactgroups' for MySQL

* In MySQL, 'groups' is a reserved keyword. We've changed it to artifactgroups,
  so one doesn't have to bother enclosing the table name with ` backticks

iobruno committed:
Add GetNextSequenceValue for mysql.ddl

* This is required for the getNextSequenceValue on MySQLSqlStatements,
  since MySQL doesn't have an equivalent to Postgres' RETURNING or
  SQL Server's OUTPUT

iobruno committed:
Drop Indexes on mysql.ddl

* Dropped IDX_versions_4 on versions as MySQL can't index VARCHAR(1024)

* Dropped IDX_props_2 on properties as MySQL can't index VARCHAR(1024)

iobruno committed:
Fix pattern to identify PK violation on MySQL

* Integrity tests identify PK violation by checking the exception err msg
  This fixes it the pattern for MySQL

iobruno committed:
Add insertVersion and autoUpdateVersionForGlobalId to MySQLSqlStatemnts

* For 'insertVersion', MySQL doesn't handle 'INSERT INTO... VALUES (..., (SELECT...), ...)' well,
  instead use 'INSERT INTO... SELECT'

* For 'autoUpdateVersionForGlobalId', MySQL doesn't behave well with
  'UPDATE... SET version = (SELECT... WHERE) WHERE...', so a CTE was necessary here

iobruno committed:
Update Makefile with entries for MySQL

iobruno committed:
Update README
